File: question/neetcode/heap/find-median-from-data-stream.cpp
#include <bits/stdc++.h>
using namespace std;

class MedianFinder {
public:
    priority_queue<int> lowerhalf;
    priority_queue<int, vector<int>, greater<int>> upperhalf;

    MedianFinder() {
        
    }
    
    void addNum(int num) {

        if(lowerhalf.empty() or upperhalf.empty()){
            lowerhalf.push(num);
        }
        else {
            if(num < lowerhalf.top()){
                lowerhalf.push(num);
            }else{
                upperhalf.push(num);
            }
        }

        if(lowerhalf.size() > upperhalf.size()+1){
            upperhalf.push(lowerhalf.top());
            lowerhalf.pop();
        }
        if(upperhalf.size() > lowerhalf.size()+1){
            lowerhalf.push(upperhalf.top());
            upperhalf.pop();
        }
    }
    
    double findMedian() {
        if(lowerhalf.size() == upperhalf.size()){
            return (double)((lowerhalf.top() + upperhalf.top())/2.0);
        }
        else if(lowerhalf.size() > upperhalf.size()){
            return lowerhalf.top();
        }else{
            return upperhalf.top();
        }
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: question/neetcode/heap/find-median-from-data-stream.py
import heapq

"""
[1,2,-1]

[1], [2]
"""


class MedianFinder:

    def __init__(self):
       self.smaller = [] 
       self.bigger = [] 

    def addNum(self, num: int) -> None:
        if not self.smaller and not self.bigger:
            heapq.heappush(self.smaller, -num)
        else:
            biggest = -self.smaller[0]

            if num <= biggest:
                heapq.heappush(self.smaller, -num)
                if len(self.smaller) > len(self.bigger) + 1:
                    n = -heapq.heappop(self.smaller)
                    heapq.heappush(self.bigger, n)
            else:
                heapq.heappush(self.bigger, num)
                if len(self.bigger) > len(self.smaller) + 1:
                    n = heapq.heappop(self.bigger)
                    heapq.heappush(self.smaller, -n)

    def findMedian(self) -> float:
        if len(self.smaller) > len(self.bigger):
            return -self.smaller[0]
        elif len(self.smaller) < len(self.bigger):
            return self.bigger[0]
        else:
            return (-self.smaller[0] + self.bigger[0])/2
        


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()